3612 Add keyboard shortcuts to gallery expanded view

jsomeara opened the pull request:
Resolves #3612

I added keyboard shortcuts to the gallery expanded view (the view when one of the gallery cards/labels is clicked).

Here are the shortcuts I added:

- Left arrow = previous label
- Right arrow = next label
- A or Y = Agree with label
- D or N = Disagree with label
- U = Unsure about label

I also made sure the keybindings for google street view (examples: W, A, S, D, arrow keys) were properly deactivated to avoid interference with the new keyboard shortcuts.

Please don't hesitate to let me know if something needs to be corrected!

##### Video

https://github.com/user-attachments/assets/f2f970fd-fbfb-4526-98fd-f871f66fcb39

##### Testing instructions
1. Go to the gallery. Click any label and this will open the expanded view.
2. Try pressing some keyboard shortcuts (the shortcuts added can be seen in the description of this PR)
3. You can also see that after exiting the expanded view, the keyboard shortcuts shouldn't perform any actions

##### Things to check before submitting the PR <!-- if something doesn't apply, just check the box or remove the line -->
<!-- You can check the box by replacing the space with an "x". So instead of "- [ ]" it would be "- [x]" -->
- [x] I've written a descriptive PR title. <!-- No need to include the issue number. Just a half sentence summary of the fix/feature! -->
- [x] I've added/updated comments for large or confusing blocks of code.

misaugstad commented on public/javascripts/Gallery/src/modal/Modal.js:
```diff
@@ -84,6 +89,26 @@ function Modal(uiModal) {
+    /**
+     * Fires when a keydown event occurs. Removed after modal closes.
+     */
+    function keydownListener(event) {
```
Can you actually create a new `Keyboard.js` file, similar to the one for SVValidate (And SVLabel)? Then I think that most of the keyboard-related stuff should go in there! Since we're finally adding shortcuts, may as well keep it consistent with our other tools!

jsomeara replied:
Is it okay if I expose the onClickOrKeyDown function and then call it from the newly created Keyboard.js file or should I move all the logic, including the http request, to the Keyboard.js file?

misaugstad replied:
I think that calling `onClickOrKeyPress` from `Keyboard.js` is best! Ideally `Keyboard.js` should be doing very little other than logging key presses and calling the appropriate functions elsewhere. Don't want too much of the logic in there :)

misaugstad commented on public/javascripts/Gallery/src/modal/Modal.js:
```diff
@@ -149,10 +175,10 @@ function Modal(uiModal) {
-            function() { sg.tracker.push('GSVInfoCopyToClipboard_Click', { panoId: getPanoId() }); },
-            function() { sg.tracker.push('GSVInfoViewInGSV_Click', { panoId: getPanoId() }); },
-            function() { return properties['label_id']; }
+            function () { sg.tracker.push('GSVInfoButton_Click', { panoId: getPanoId() }); },
```
Let's keep no space between these! I think that's what we have in the style guide?

jsomeara replied:
Sorry about this! My auto formatter changed it and I didn't notice.

misaugstad commented on public/javascripts/Gallery/src/validation/ValidationMenu.js:
```diff
@@ -107,9 +140,10 @@ function ValidationMenu(refCard, gsvImage, cardProperties, modal, onExpandedView
-    function validateOnClick(newValKey, thumbsClick) {
-        return function(e) {
-            if (currSelected !== newValKey) {
+    function validateOnClickOrKeyPress(newValKey, thumbsClick) {
```
Can you also make sure that we log to `sg.tracker` differently for clicks vs shortcuts?

jsomeara replied:
![image](https://github.com/user-attachments/assets/afff4c7b-655f-4b3a-a108-d7b5d9865306)
![image](https://github.com/user-attachments/assets/16976455-51f0-401d-80f7-870d1e591d9e)

Would these be the areas I need to change? Do I need to do anything else other than just replace the "Click" with "KeyDown"?

misaugstad replied:
I think the only other place is when moving to the next label with arrow keys, but it isn't moving to an entirely new page. I'm assuming that we are logging that... If we aren't, then we should, and now would be a good time to add that!

Only other caveat is that I think we typically use "KeyboardShortcut" rather than "KeyDown" in the action names!

jsomeara replied:
Got it! Sorry, I confused the NextPage/PrevPage as being the next/prev label for some reason.

I'll make sure to use `KeyboardShortcut_`

I could be completely wrong here, but I can't seem to find any logging for the next/previous labels. I'll create the proper logging for those (both the click version and the KeyboardShortcut version).

Do I need to update some documentation anywhere after adding this logging? I know there's [this](https://github.com/ProjectSidewalk/SidewalkWebpage/wiki/Descriptions-of-Logged-Events).

misaugstad replied:
>I could be completely wrong here, but I can't seem to find any logging for the next/previous labels. I'll create the proper logging for those (both the click version and the KeyboardShortcut version).

That wouldn't surprise me! The logging in Gallery is lacking compared to Explore/Validate.

>Do I need to update some documentation anywhere after adding this logging? I know there's [this](https://github.com/ProjectSidewalk/SidewalkWebpage/wiki/Descriptions-of-Logged-Events).

That would be the place to update it, yes! Would need to add a new Gallery column. That page tends to get a bit out of date, but I would encourage every effort to keep it updated!!

misaugstad reviewed (changes requested):
Thanks for getting started on this!! I added a few notes on restructuring to do to be consistent with other parts of our code!

jsomeara commented:
I made a dedicated file for the keyboard event listening called `Keyboard.js`. I also fixed the inline function declarations to match the style guide. Lastly, I added all of the logging necessary (I am yet to update the wiki page yet).

misaugstad commented on public/javascripts/Gallery/src/keyboard/Keyboard.js:
```diff
@@ -0,0 +1,51 @@
+     * @param code event.code to match for
+     * @param action Function to execute when the event.code is matched
+     */
+    function bindCodeToAction(code, action) {
```
Hmm, I'm not crazy about Keyboard.js being hands off and letting anywhere else bind the actions to key events. What happens if we decide that the left/right arrow keys should have a different meaning when the expanded view is closed? If another part of the code binds an action to the same key, it just overwrites the one that already exists.

I'd rather have all the keyboard related stuff in Keyboard.js so that it can deal with potentially more complex logic in the future.

jsomeara replied:
I agree. Not only is it better for long-term changes, it also is more consistent with the other Keyboard.js files. I'll make some changes.

misaugstad commented on public/javascripts/Gallery/src/cards/CardContainer.js:
```diff
@@ -92,6 +92,7 @@ function CardContainer(uiCardContainer, initialFilters) {
         sg.ui.cardContainer.holder.on('click', '.static-gallery-image, .additional-count',  (event) => {
             $('.gallery-modal').attr('style', 'display: flex');
             $('.grid-container').css("grid-template-columns", "1fr 5fr");
+            modal.open = true;
```
Is there a reason that this is being done here instead of just in the `openModal()` function?

jsomeara replied:
Sorry about that! For some reason I didn't see that function.

misaugstad commented on public/javascripts/Gallery/src/modal/Modal.js:
```diff
@@ -24,6 +24,8 @@ function Modal(uiModal) {
                 $('.gallery-modal').attr('style', 'display: flex');
                 $('.grid-container').css("grid-template-columns", "1fr 5fr");
 
+                self.open = true;
```
And if we can include set this in `openModal()`, we shouldn't need to do it here. But maybe you're doing it before `openModal()` is called for a reason?

misaugstad commented on public/javascripts/Gallery/src/validation/ValidationMenu.js:
```diff
@@ -97,26 +105,28 @@ function ValidationMenu(refCard, gsvImage, cardProperties, modal, onExpandedView
+    function validateOnClickOrKeyPress(newValKey, thumbsClick, keyboardShortcut) {
         return function(e) {
-            if (currSelected !== newValKey) {
+            // If we aren't just doing what's already been selected, we have the card properties, and modal is open (this last predicate is only necessary if this is the validation menu for the expanded view).
```
Make sure to keep lines under 120 characters!

misaugstad reviewed (changes requested):
Okay there's some restructuring I'm looking for throughout the code, mostly to be better prepared for the future! But also there are some issues like being able to hold down an arrow key and fast forward through all the labels on a page without any of them rendering :)

Though it did work great and I love a lot of what you've done here!

jsomeara commented on public/javascripts/Gallery/src/keyboard/Keyboard.js:
```diff
@@ -0,0 +1,51 @@
+            // Execute the bound action.
+            self.codeToAction[e.code]();
+        }
+    };
```
Ah! Forgot to remove those when moving from arrow functions to regular functions.

jsomeara commented on public/javascripts/Gallery/src/keyboard/Keyboard.js:
```diff
@@ -0,0 +1,51 @@
+     * @param {object} e An event object
+     * @private
+     */
+    function _documentKeyDown(e) {
```
Definitely, agreed.

jsomeara commented:
I made some changes to address the issues. The biggest one was restructuring the way the system worked. Instead of having a keybinding function, it centralizes the keybindings in the Keyboard.js file by using a switch statement.

misaugstad commented:
>Sorry for having that many commit messages! I didn't realize how many times I was committing even minor changes.

No problem! I'd rather have commit messages that are two small over ones that are too large! And that line can be harder to find when you're making a bunch of misc changes from a PR review.

Hoping to take a look later today or tomorrow! :crossed_fingers:

jsomeara commented:
Implemented the requested changes! I also caught an issue where it was unnecessarily spitting out error messages in the console when e.key was undefined. That's fixed now. Ready for review!

misaugstad reviewed (approved):
Perfect, so happy to have these shortcuts in Gallery finally!! Thanks @jsomeara !
